lib/model/message.dart

```diff
@@ -167,33 +168,61 @@ class MessageStoreImpl with MessageStore {
       return;
     }
 
-    if (newTopic == null) {
-      // The `subject` field (aka newTopic) is documented to be present on moves.
-      assert(debugLog('Malformed UpdateMessageEvent: move but no newTopic')); // TODO(log)
+    if (newTopic == null && newStreamId == null) {
+      // At least one of `subject` (aka newTopic) or `new_stream_id` (aka
+      // newStreamId) is documented to be present on moves.
+      assert(debugLog('Malformed UpdateMessageEvent: move but no newTopic or newStreamId')); // TODO(log)
       return;
     }
     if (origStreamId == null) {
       // The `stream_id` field (aka origStreamId) is documented to be present on moves.
       assert(debugLog('Malformed UpdateMessageEvent: move but no origStreamId')); // TODO(log)
       return;
     }
-
-    if (newStreamId == null
-        && MessageEditState.topicMoveWasResolveOrUnresolve(origTopic, newTopic)) {
-      // The topic was only resolved/unresolved.
-      // No change to the messages' editState.
+    if (propagateMode == null) {
+      // The `propagate_mode` field (aka propagateMode) is documented to be present on moves.
+      assert(debugLog('Malformed UpdateMessageEvent: move but no propagateMode')); // TODO(log)
       return;
     }
 
-    // TODO(#150): Handle message moves.  The views' recipient headers
-    //   may need updating, and consequently showSender too.
-    //   Currently only editState gets updated.
+    final wasResolveOrUnresolve = (newStreamId == null && newTopic != null
+      && MessageEditState.topicMoveWasResolveOrUnresolve(origTopic, newTopic));
+
     for (final messageId in event.messageIds) {
       final message = messages[messageId];
       if (message == null) continue;
-      // Do not override the edited marker if the message has also been moved.
-      if (message.editState == MessageEditState.edited) continue;
-      message.editState = MessageEditState.moved;
+
+      if (message is! StreamMessage) {
+        assert(debugLog('Bad UpdateMessageEvent: stream/topic move on a DM')); // TODO(log)
+        continue;
+      }
+
+      if (newStreamId != null) {
+        message.streamId = newStreamId;
+        // See [StreamMessage.displayRecipient] on why the invalidation is
+        // needed.
+        message.displayRecipient = null;
+      }
+
+      if (newTopic != null) {
+        message.topic = newTopic;
+      }
+
+      if (!wasResolveOrUnresolve
+          && message.editState == MessageEditState.none) {
+        message.editState = MessageEditState.moved;
+      }
+    }
+
+    for (final view in _messageListViews) {
+      view.messagesMoved(
```

Comment:
nit in commit message:

    Signed-off-by: Zixuan James Li <zixuan@zulip.com>
    Co-authored-by: Zixuan James Li <zixuan@zulip.com>

If we're going to have "Signed-off-by:" lines, let's follow the conventions of how they're used. It's a trailer that comes after the material you're signing off on. Are you signing off on the parts of this commit which you co-authored, or only on other parts? If the former, your "Signed-off-by:" should come after your "Co-authored-by:". (And if the latter, it's not clear what the line could really mean and it's probably best to just leave it out.)

For example usage, go into the Linux kernel repo and try a command like `git log --grep Co-author`.